### PR TITLE
feat(otlphttp): configure response body size limit

### LINKED ISSUE #8819: feat(otlphttp): configure response body size limit
### Problem Statement

The OTLP/HTTP specification requires clients to limit response-body size and recommends making that limit configurable. The Go OTLP/HTTP trace, metric, and log exporters currently use a fixed 4 MiB limit, so users cannot lower the limit to reduce memory exposure or raise it for legitimate larger responses.

This work was started in #8443. The original author no longer has bandwidth to continue that PR, so this issue tracks the remaining work for another contributor.

### Proposed Solution

Add a `WithMaxResponseBodySize(size int64)` option to the OTLP/HTTP trace, metric, and log exporters.

- Preserve the current finite 4 MiB default.
- Apply the configured limit to response-body bytes read after decompression.
- Do not allow non-positive values to disable the limit; reject them or retain the default.
- Keep the shared configuration templates and generated exporter code consistent.
- Add tests and documentation for the default, configured, and invalid-value behavior.

#### Alternatives

Keep the fixed 4 MiB limit. This preserves the existing safety bound but does not implement the specification recommendation that the limit be configurable.

#### Prior Art

- Specification change: open-telemetry/opentelemetry-proto#802
- Initial implementation and review discussion: #8443
- Safety review about retaining a finite positive limit: https://github.com/open-telemetry/opentelemetry-go/pull/8443#discussion_r3553257280

### Additional Context

The OTLP/HTTP specification describes the response-size requirement here: https://github.com/open-telemetry/opentelemetry-proto/blob/41ad255e3efd5450096fc167ef6d87edb4bcb4f3/docs/specification.md#L522-L527

<sub>**Tip**: [React](https://github.blog/news-insights/product-news/add-reactions-to-pull-requests-issues-and-comments/) with 👍 to help prioritize this issue. Please use comments to provide useful context, avoiding `+1` or `me too`, to help us triage it. Learn more [here](https://opentelemetry.io/community/end-user/issue-participation/).</sub>

### PR DESCRIPTION
**Pull request by jstar0**
## Summary

Fixes #8819.

- Add `WithMaxResponseBodySize` to the OTLP/HTTP trace, metric, and log exporters.
- Keep the finite 4 MiB default response-body limit and apply it after decompression.
- Treat non-positive option values as invalid and retain the default instead of disabling the safety limit.
- Use bounded reads that detect oversized responses without integer-overflow edge cases.
- Keep the generated gRPC configuration free of HTTP-only response-body settings.

## Verification

- `go test ./...` passes in the trace, metric, and log OTLP/HTTP exporter modules.
- `make precommit` completed generation, formatting/lint, README validation, and module consistency checks.
- The repository-wide race run reached the pre-existing `bridge/opentracing` gRPC test, which failed locally with an EOF while reading the server preface; no changed package failed.

### CODE REVIEW

File: exporters/otlp/otlptrace/otlptracehttp/client.go [hunk 1]
```diff
@@ -402,6 +388,35 @@ func (r *request) reset(ctx context.Context) {
...
+type responseBodyTooLargeError struct {
+	limit int64
+}
+
+func (e responseBodyTooLargeError) Error() string {
+	return fmt.Sprintf("response body too large: exceeded %d bytes", e.limit)
+}
+
+func copyResponseBody(dst io.Writer, src io.ReadCloser, maxSize int64) error {
+	if maxSize <= 0 || maxSize == math.MaxInt64 {
```
Comment (MrAlias) [inline exporters/otlp/otlptrace/otlptracehttp/client.go]: Keep `MaxInt64` on the bounded path. `WithMaxResponseBodySize(math.MaxInt64)` reaches this branch and changes the read to an unbounded `io.Copy`, despite the option's guarantee that the response-body limit cannot be disabled and issue #8819's requirement that every configuration retain a finite positive bound. A server can then keep the response open and stream data until the host runs out of memory without this helper ever returning `responseBodyTooLargeError`. The following `LimitReader(maxSize)` plus one-byte probe does not add to `maxSize`, so it has no integer-overflow problem at `MaxInt64`; please remove this special case in trace, metric, and logs.

File: CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,7 @@ The next release will require at least [Go 1.26].
 
 ### Added
 
+- Add configurable response body size limits to the OTLP/HTTP trace, metric, and log exporters. (#8847)
```
Comment (MrAlias) [inline CHANGELOG.md]: Use the affected module names in this entry. Repository changelog guidance requires user-visible entries to identify updated Go modules rather than only package categories. This should name `go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp`, `go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp`, and `go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploghttp`.

**Review by MrAlias — COMMENTED** [review]
The PR adds the requested option consistently across trace, metric, and log OTLP/HTTP exporters, preserves the 4 MiB default for ordinary configurations, updates generated configuration, and adds focused tests.

**Comment by jstar0** [comment]
Heads-up while refreshing this branch: current `main` now ships its own response-body limit (package-level `maxResponseBodySize` var, enforced via `http.MaxBytesReader` with `MaxBytesError` handling, from the changes around #8876). That limit is not user-configurable — the variable exists so tests can override it — while this PR adds `WithMaxResponseBodySize` as a per-exporter option.

The two overlap, so before I rebase I'd like to align on the intended shape:

1. **Compose them (my preference):** keep `main`'s package var as the default source, and let `WithMaxResponseBodySize` override it per exporter. This preserves `TestResponseBodySizeLimit`'s var-override pattern unchanged and adds the configurability this PR (and #8819) ask for.
2. **Var-only:** if maintainers prefer the non-configurable cap as final, this PR becomes redundant for the option parts and I can close it in favor of `main`'s implementation.

Happy to rebase under either decision — the option plumbing (oconf + trace/metric/log exporters) is already in place on this branch.

**Comment by jstar0** [comment]
Merged current `main` and composed the two limits as described above:

- Package-level `maxResponseBodySize` remains the default (tests can still override it).
- `WithMaxResponseBodySize` overrides that default per exporter.
- Every response-body read, including `WithMaxResponseBodySize(math.MaxInt64)`, goes through `http.MaxBytesReader`.
- Changelog entry now names the three HTTP exporter modules.

Verified with `go test` in `otlptracehttp`, `otlpmetrichttp`, `otlploghttp`, plus their `otlpconfig`/`oconf` packages.

File: exporters/otlp/otlptrace/otlptracehttp/client.go (hunk 1 above)
Comment (jstar0) [reply exporters/otlp/otlptrace/otlptracehttp/client.go]: Addressed in the merge onto current `main`. `copyResponseBody` always uses `http.MaxBytesReader`; `maxSize <= 0` falls back to the package-level 4 MiB default, and `math.MaxInt64` stays on that bounded path. Added `TestCopyResponseBodyMaxInt64StaysBounded` in the trace, metric, and log HTTP clients.

File: CHANGELOG.md (hunk 2 above)
Comment (jstar0) [reply CHANGELOG.md]: Moved the Unreleased entry to name the three modules: `go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp`, `go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp`, and `go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploghttp`.

File: internal/shared/otlp/otlptrace/otlpconfig/options.go.tmpl [hunk 3]
```diff
@@ -82,13 +97,19 @@ type (
 func NewHTTPConfig(opts ...HTTPOption) Config {
 	cfg := Config{
 		Traces: SignalConfig{
-			Endpoint:       fmt.Sprintf("%s:%d", DefaultCollectorHost, DefaultCollectorHTTPPort),
+{{ if .isHTTP }}			Endpoint:       fmt.Sprintf("%s:%d", DefaultCollectorHost, DefaultCollectorHTTPPort),
```
Comment (dashpole) [inline internal/shared/otlp/otlptrace/otlpconfig/options.go.tmpl]: Both branches of this `{{ if .isHTTP }}` check look identical now that `MaxResponseBodySize` isn't set here. We can drop the conditional and keep a single struct literal.

File: internal/shared/otlp/otlptrace/otlpconfig/options.go.tmpl [hunk 4]
```diff
@@ -34,6 +34,11 @@ const (
 	// DefaultMaxRequestSize is the default maximum size of a serialized export
 	// request, before compression.
 	DefaultMaxRequestSize int = 64 * 1024 * 1024
+{{ if .isHTTP }}
+	// DefaultMaxResponseBodySize is the default maximum size of an OTLP/HTTP
+	// response body, including after decompression.
+	DefaultMaxResponseBodySize int64 = 4 * 1024 * 1024
```
Comment (dashpole) [inline internal/shared/otlp/otlptrace/otlpconfig/options.go.tmpl]: `DefaultMaxResponseBodySize` does not seem to be referenced anywhere since the default is now handled via `maxResponseBodySize` in `client.go`. Can we remove it?

File: exporters/otlp/otlplog/otlploghttp/config.go [hunk 5]
```diff
@@ -23,12 +23,13 @@ import (
...
-	defaultPath                                  = "/v1/logs"
-	defaultTimeout                               = 10 * time.Second
-	defaultMaxRequestSize                        = 64 * 1024 * 1024
-	defaultProxy          HTTPTransportProxyFunc = http.ProxyFromEnvironment
-	defaultRetryCfg                              = retry.DefaultConfig
+	defaultEndpoint                                   = "localhost:4318"
+	defaultPath                                       = "/v1/logs"
+	defaultTimeout                                    = 10 * time.Second
+	defaultMaxRequestSize                             = 64 * 1024 * 1024
+	defaultMaxResponseBodySize                        = int64(4 * 1024 * 1024)
```
Comment (dashpole) [inline exporters/otlp/otlplog/otlploghttp/config.go]: `defaultMaxResponseBodySize` is not used in `newConfig`, which resolves using `maxResponseBodySize` from `client.go` instead. We could clean this up by removing the unused variable or using it as the fallback in `newConfig`.

**Comment by jstar0** [comment]
Closing this in favor of #8941, which landed the same capability with the `WithMaxResponseSize` spelling aligned with declarative configuration and the shared template generation. Thanks @MrAlias for the reviews on both efforts — no need to keep the duplicate open.
